src/software/estop/threaded_estop_reader.cpp

```diff
@@ -0,0 +1,103 @@
+
+#include "threaded_estop_reader.h"
+
+#include <boost/bind/bind.hpp>
+#include <thread>
+#include <utility>
+
+#include "software/logger/logger.h"
+
+
+ThreadedEstopReader::ThreadedEstopReader(std::unique_ptr<UartCommunication> uart_reader,
+                                         unsigned int startup_time_ms,
+                                         unsigned int regular_interval_time_ms)
+    : estop_state(EstopState::STOP),
+      regular_read_interval_ms(regular_interval_time_ms),
+      timer(io_service, boost::posix_time::milliseconds(startup_time_ms)),
+      uart_reader(std::move(uart_reader))
+{
+    estop_thread = std::thread(boost::bind(&ThreadedEstopReader::continousRead, this));
+}
+
+void ThreadedEstopReader::continousRead()
+{
+    timer.async_wait(
+        boost::bind(&ThreadedEstopReader::tick, this, boost::asio::placeholders::error));
+    io_service.run();
+}
+
+bool ThreadedEstopReader::isEstopPlay()
+{
+    return estop_state == EstopState::PLAY;
+}
+
+void ThreadedEstopReader::tick(const boost::system::error_code& error)
+{
+    if (in_destructor)
+    {
+        timer.cancel();
+    }
+    else
+    {
+        std::vector<unsigned char> estop_msg;
+        try
+        {
+            uart_reader->flushSerialPort(uart_reader->flush_receive);
+            estop_msg = uart_reader->serialRead(ESTOP_MESSAGE_SIZE_BYTES);
+        }
+        catch (std::exception e)
+        {
+            LOG(FATAL) << "ESTOP reader failure: " << e.what();
+        }
+
+        unsigned int next_interval_ms = regular_read_interval_ms;
+        EstopState new_state;
+
+        switch (estop_msg.at(0))
+        {
+            case ESTOP_PLAY:
+                new_state                    = EstopState::PLAY;
+                num_consecutive_status_error = 0;
+                break;
+            case ESTOP_STOP:
+                new_state                    = EstopState::STOP;
+                num_consecutive_status_error = 0;
+                break;
+            default:
+                new_state = EstopState::STATUS_ERROR;
+                LOG(WARNING) << "read unexpected estop message";
+                num_consecutive_status_error++;
+                next_interval_ms = INTERVAL_BETWEEN_ERROR_READS_MS;
```

Comment:
normal interval is passed in as a member, so this acts like protection in case the normal interval is too long, since we prefer to determine quickly if the unknown message was just a mishap or an actual error

Reply:
ah I see that makes sense, in that case we can probably remove it from the constructor and just make this the `INTERVAL_BETWEEN_READS_MS` and set it in the class as a const variable

we wouldn't want the user to configure this normal interval to be very long, they can always change the interval in the class if needed

Reply:
removed from constructor and set INTERVAL_BETWEEN_READS_MS as const variable